module_utils/foreman_helper.py

```diff
@@ -51,22 +53,193 @@ def filter_module_params(self):
+                    fields.append(key)
+            if force_update:
+                for key in force_update:
+                    fields.append(key)
+            if check_missing is not None:
+                for key in check_missing:
+                    if key in entity_dict and key not in volatile_entity.keys():
+                        volatile_entity[key] = new_value
+                        fields.append(key)
+            if len(fields) > 0:
+                if not self.check_mode:
+                    new_data = {}
+                    for key, value in volatile_entity.items():
+                        if key in fields:
+                            new_data[key] = value
+                    result = self.foremanapi.resource(resource).call('update', self._generate_resource_params(resource, 'update', resource_id=entity_id, params=new_data))
+                return True, result
+            return False, result
+        except Exception as e:
+            self.fail_json(msg='Error while updating {0}: {1}'.format(
+                resource, str(e)))
+
+    def ensure_resource_state(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
+        changed, _ = self.ensure_resource(resource, entity_dict, entity, state, check_missing, check_type, force_update)
+        return changed
+
+    def ensure_resource(self, resource, entity_dict, entity, state, check_missing=None, check_type=None, force_update=None):
+        """ Ensure that a given entity has a certain state """
+        changed, changed_entity = False, entity
+
+        if state == 'present':
+            if entity is None:
+                changed, changed_entity = self.create_resource(resource, entity_dict)
+            else:
+                changed, changed_entity = self.update_resource(resource, entity, entity_dict, check_missing, check_type, force_update)
+        elif state == 'absent':
+            if entity is not None:
+                changed, changed_entity = self.delete_resource(resource, entity['id'])
+        else:
+            self.fail_json(msg='Not a valid state: {}'.format(state))
+        return changed, changed_entity
+
+    def _resource_action(self, resource, action, params):
+        try:
+            result = None
+            if not self.check_mode:
+                result = self.foremanapi.resource(resource).call(action, params)
+        except Exception as e:
+            self.fail_json(msg='Error while {}ing {}: {}'.format(
+                action, resource, str(e)))
+        return True, result
+
+    def _generate_resource_params(self, resource, action, resource_id = None, params = None):
+        resource_params = {}
+        if resource_id is not None:
+            resource_params['id'] = resource_id
+        if params is not None:
+            action_params = self.foremanapi.resource(resource).action(action).params
+            hash_params = [p.name for p in action_params if p.expected_type == 'hash']
```

Comment:
@sean797 welp, this works, is it magic?

Reply:
I think that will work for Foreman but not every API expects a hash, especially Katello, e.g https://theforeman.org/plugins/katello/3.9/api/apidoc/v2/content_views/create.html